File: app/src/main/java/pl/karol202/paintplus/tool/selection/ToolSelection.java
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package pl.karol202.paintplus.tool.selection;

import android.graphics.*;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.image.LegacyImage;
import pl.karol202.paintplus.tool.StandardTool;
import pl.karol202.paintplus.tool.ToolCoordinateSpace;
import pl.karol202.paintplus.tool.ToolProperties;

import static pl.karol202.paintplus.tool.selection.ToolSelectionShape.OVAL;
import static pl.karol202.paintplus.tool.selection.ToolSelectionShape.RECTANGLE;

public class ToolSelection extends StandardTool
{
	private enum MoveType
	{
		NONE,
		LEFT_TOP_CORNER, RIGHT_TOP_CORNER, LEFT_BOTTOM_CORNER, RIGHT_BOTTOM_CORNER,
		LEFT_SIDE, TOP_SIDE, RIGHT_SIDE, BOTTOM_SIDE,
		MOVE
	}

	private static final int MAX_DISTANCE_DP = 50;
	private static final float SELECTION_LINE_WIDTH_DP = 1f;

	private final int MAX_DISTANCE_PX;
	private final float SELECTION_LINE_WIDTH_PX;

	private ToolSelectionShape shape;
	private ToolSelectionMode mode;

	private OnSelectionEditListener selectionListener;
	private Rect rect;
	private boolean editMode;
	private boolean rectCreated;
	private Paint paint;
	private Rect dirtyRect;

	private MoveType movingType;
	private Rect rectAtBeginning;
	private Point movingStart;

	public ToolSelection(LegacyImage image)
	{
		super(image);
		MAX_DISTANCE_PX = (int) (MAX_DISTANCE_DP * image.SCREEN_DENSITY);
		SELECTION_LINE_WIDTH_PX = (int) (SELECTION_LINE_WIDTH_DP * image.SCREEN_DENSITY);

		this.shape = RECTANGLE;
		this.mode = ToolSelectionMode.NEW;

		this.rect = new Rect();
		this.paint = new Paint();
		this.paint.setAntiAlias(true);
		this.paint.setColor(Color.BLACK);
		this.paint.setStyle(Paint.Style.STROKE);

		cleanUp();
	}

	@Override
	public int getName()
	{
		return R.string.tool_selection;
	}

	@Override
	public int getIcon()
	{
		return R.drawable.ic_tool_selection_black_24dp;
	}

	@Override
	public Class<? extends ToolProperties> getPropertiesFragmentClass()
	{
		return SelectionProperties.class;
	}

	@Override
	public ToolCoordinateSpace getInputCoordinateSpace()
	{
		return ToolCoordinateSpace.IMAGE_SPACE;
	}

	@Override
	public boolean isUsingSnapping()
	{
		return false;
	}

	@Override
	public boolean onTouchStart(float xFloat, float yFloat)
	{
		if(image.getSelectedLayer() == null)
		{
			if(editMode) cleanUp();
			return false;
		}

		int x = Math.round(xFloat);
		int y = Math.round(yFloat);

		if(!editMode) enableEditMode();
		if(!rectCreated)
		{
			setLeft(x);
			setTop(y);
		}
		else
		{
			setMovingType(x, y);
			rectAtBeginning = new Rect(rect);
			movingStart = new Point(x, y);
		}
		expandDirtyRect();

		return true;
	}

	private void enableEditMode()
	{
		editMode = true;
		if(selectionListener != null) selectionListener.onStartSelectionEditing();
	}

	private void setMovingType(int x, int y)
	{
		int leftDist = Math.abs(rect.left - x);
		int topDist = Math.abs(rect.top - y);
		int rightDist = Math.abs(rect.right - x);
		int bottomDist = Math.abs(rect.bottom - y);

		boolean left = leftDist < (MAX_DISTANCE_PX / image.getZoom());
		boolean top = topDist < (MAX_DISTANCE_PX / image.getZoom());
		boolean right = rightDist < (MAX_DISTANCE_PX / image.getZoom());
		boolean bottom = bottomDist < (MAX_DISTANCE_PX / image.getZoom());
		boolean xInside = x > rect.left && x < rect.right;
		boolean yInside = y > rect.top && y < rect.bottom;

		movingType = MoveType.NONE;

		if(xInside && yInside) movingType = MoveType.MOVE;

		if(left && !right && yInside) movingType = MoveType.LEFT_SIDE;
		if(top && !bottom && xInside) movingType = MoveType.TOP_SIDE;
		if(right && !left && yInside) movingType = MoveType.RIGHT_SIDE;
		if(bottom && !top && xInside) movingType = MoveType.BOTTOM_SIDE;

		if(left && top) movingType = MoveType.LEFT_TOP_CORNER;
		if(right && top) movingType = MoveType.RIGHT_TOP_CORNER;
		if(right && bottom) movingType = MoveType.RIGHT_BOTTOM_CORNER;
		if(left && bottom) movingType = MoveType.LEFT_BOTTOM_CORNER;
	}

	@Override
	public boolean onTouchMove(float xFloat, float yFloat)
	{
		int x = Math.round(xFloat);
		int y = Math.round(yFloat);

		if(!rectCreated)
		{
			setRight(x);
			setBottom(y);
		}
		else move(x, y);
		expandDirtyRect();

		return true;
	}

	@Override
	public boolean onTouchStop(float xFloat, float yFloat)
	{
		int x = Math.round(xFloat);
		int y = Math.round(yFloat);

		if(!rectCreated)
		{
			setRight(x);
			setBottom(y);
		}
		else move(x, y);
		correctBounds();
		rectCreated = true;
		dirtyRect = null;

		return true;
	}

	private void move(int x, int y)
	{
		int deltaX = x - movingStart.x;
		int deltaY = y - movingStart.y;
		rect = new Rect(rectAtBeginning);

		switch(movingType)
		{
		case NONE:
			break;
		case LEFT_TOP_CORNER:
			setLeft(rect.left + deltaX);
			setTop(rect.top + deltaY);
			break;
		case RIGHT_TOP_CORNER:
			setRight(rect.right + deltaX);
			setTop(rect.top + deltaY);
			break;
		case LEFT_BOTTOM_CORNER:
			setLeft(rect.left + deltaX);
			setBottom(rect.bottom + deltaY);
			break;
		case RIGHT_BOTTOM_CORNER:
			setRight(rect.right + deltaX);
			setBottom(rect.bottom + deltaY);
			break;
		case LEFT_SIDE:
			setLeft(rect.left + deltaX);
			break;
		case TOP_SIDE:
			setTop(rect.top + deltaY);
			break;
		case RIGHT_SIDE:
			setRight(rect.right + deltaX);
			break;
		case BOTTOM_SIDE:
			setBottom(rect.bottom + deltaY);
			break;
		case MOVE:
			PointF oldCenter = new PointF(rect.centerX(), rect.centerY());
			PointF newCenter = new PointF(oldCenter.x + deltaX, oldCenter.y + deltaY);
			helpersManager.snapPoint(newCenter);
			rect.offset((int) (newCenter.x - oldCenter.x), (int) (newCenter.y - oldCenter.y));
			break;
		}
	}

	private void setLeft(int left)
	{
		rect.left = (int) helpersManager.snapX(left);
	}

	private void setTop(int top)
	{
		rect.top = (int) helpersManager.snapY(top);
	}

	private void setRight(int right)
	{
		rect.right = (int) helpersManager.snapX(right);
	}

	private void setBottom(int bottom)
	{
		rect.bottom = (int) helpersManager.snapY(bottom);
	}

	private void correctBounds()
	{
		if(rect.left > rect.right)
		{
			int temp = rect.left;
			rect.left = rect.right;
			rect.right = temp;
		}
		if(rect.top > rect.bottom)
		{
			int temp = rect.top;
			rect.top = rect.bottom;
			rect.bottom = temp;
		}
	}

	private void expandDirtyRect()
	{
		if(dirtyRect == null) dirtyRect = new Rect();
		dirtyRect.left = Math.min(dirtyRect.left, rect.left);
		dirtyRect.top = Math.min(dirtyRect.top, rect.top);
		dirtyRect.right = Math.max(dirtyRect.right, rect.right);
		dirtyRect.bottom = Math.max(dirtyRect.bottom, rect.bottom);
	}

	void applySelection()
	{
		if(shape == RECTANGLE) selection.commitSelectionRectangle(rect, mode.getOp());
		else if(shape == OVAL) selection.commitSelectionOval(rect, mode.getOp());
		cancelSelection();
	}

	void cancelSelection()
	{
		cleanUp();
		image.updateImage();
	}

	private void cleanUp()
	{
		editMode = false;
		rectCreated = false;
		rect.left = -1;
		rect.top = -1;
		rect.right = -1;
		rect.bottom = -1;
	}

	void selectAll()
	{
		selection.selectAll();
	}

	void selectNothing()
	{
		selection.selectNothing();
	}

	void invertSelection()
	{
		selection.revert();
	}

	@Override
	public boolean providesDirtyRegion()
	{
		return true;
	}

	@Override
	public Rect getDirtyRegion()
	{
		return dirtyRect;
	}

	@Override
	public void resetDirtyRegion()
	{
		if(dirtyRect != null) dirtyRect.setEmpty();
	}

	@Override
	public boolean doesOnLayerDraw(boolean layerVisible)
	{
		return false;
	}

	@Override
	public boolean doesOnTopDraw()
	{
		return true;
	}

	@Override
	public ToolCoordinateSpace getOnLayerDrawingCoordinateSpace()
	{
		return null;
	}

	@Override
	public ToolCoordinateSpace getOnTopDrawingCoordinateSpace()
	{
		return ToolCoordinateSpace.IMAGE_SPACE;
	}

	@Override
	public void drawOnLayer(Canvas canvas) { }

	@Override
	public void drawOnTop(Canvas canvas)
	{
		if(rect.left == -1 || rect.top == -1 || rect.right == -1 || rect.bottom == -1) return;
		paint.setStrokeWidth(SELECTION_LINE_WIDTH_PX / image.getZoom());

		if(shape == RECTANGLE) canvas.drawRect(rect, paint);
		else if(shape == OVAL) canvas.drawOval(new RectF(rect), paint);
	}

	void setSelectionListener(OnSelectionEditListener selectionListener)
	{
		this.selectionListener = selectionListener;
	}

	boolean isInEditMode()
	{
		return editMode;
	}

	ToolSelectionShape getShape()
	{
		return shape;
	}

	void setShape(ToolSelectionShape shape)
	{
		this.shape = shape;
	}

	ToolSelectionMode getMode()
	{
		return mode;
	}

	void setMode(ToolSelectionMode mode)
	{
		this.mode = mode;
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/selection/ToolSelection.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.selection

import android.content.Context
import android.graphics.*
import androidx.core.graphics.*
import pl.karol202.paintplus.R
import pl.karol202.paintplus.helpers.HelpersService
import pl.karol202.paintplus.history.Action
import pl.karol202.paintplus.image.*
import pl.karol202.paintplus.image.layer.Layer
import pl.karol202.paintplus.options.OptionSelect
import pl.karol202.paintplus.tool.StandardTool
import pl.karol202.paintplus.tool.ToolCoordinateSpace
import pl.karol202.paintplus.util.*
import pl.karol202.paintplus.util.MathUtils.dpToPixels
import kotlin.math.abs
import kotlin.math.roundToInt

private const val MAX_DRAG_DISTANCE_DP = 50f
private const val SELECTION_LINE_WIDTH_DP = 1f

class ToolSelection(context: Context,
                    private val imageService: ImageService,
                    private val viewService: ViewService,
                    effectsService: EffectsService,
                    private val helpersService: HelpersService,
                    private val historyService: HistoryService,
                    private val optionSelect: OptionSelect) :
		StandardTool(imageService, viewService, helpersService, effectsService)
{
	private enum class MoveType
	{
		NONE,
		LEFT_TOP_CORNER, RIGHT_TOP_CORNER, LEFT_BOTTOM_CORNER, RIGHT_BOTTOM_CORNER,
		LEFT_SIDE, TOP_SIDE, RIGHT_SIDE, BOTTOM_SIDE,
		MOVE
	}

	private sealed class State
	{
		object Idle : State()

		sealed class Edit : State()
		{
			data class Initial(override val rect: Rect) : Edit()

			data class Adjust(override val rect: Rect,
			                  val rectAtStart: Rect,
			                  val moveType: MoveType,
			                  val moveStart: Point) : Edit()

			abstract val rect: Rect
		}
	}

	private val maxDragDistancePx = dpToPixels(context, MAX_DRAG_DISTANCE_DP)
	private val selectionLineWidthPx = dpToPixels(context, SELECTION_LINE_WIDTH_DP)

	override val name get() = R.string.tool_selection
	override val icon get() = R.drawable.ic_tool_selection_black_24dp
	override val propertiesFragmentClass get() = SelectionProperties::class.java
	override val inputCoordinateSpace get() = ToolCoordinateSpace.IMAGE_SPACE
	override val isUsingSnapping get() = false

	var shape = ToolSelectionShape.RECTANGLE
	var mode = ToolSelectionMode.NEW

	private val actionPreset = Action.namePreset(R.string.history_action_selection_change).withPreview {
		optionSelect.createPreviewBitmap()
	}
	private val paint = Paint().apply {
		isAntiAlias = true
		color = Color.BLACK
		style = Paint.Style.STROKE
	}

	private var state by notifying<State>(State.Idle)
	val isInEditMode get() = state is State.Edit

	override fun onTouchStart(point: PointF, layer: Layer): Boolean
	{
		state = getInitialState(point.toPoint())
		return true
	}

	private fun getInitialState(point: Point) = when(val state = state)
	{
		is State.Idle -> State.Edit.Initial(Rect(snapX(point.x), snapY(point.y),
		                                         snapX(point.x), snapY(point.y)))
		is State.Edit -> State.Edit.Adjust(state.rect, state.rect, getMoveType(point, state.rect), point)
	}

	private fun getMoveType(point: Point, rect: Rect): MoveType
	{
		val x = point.x
		val y = point.y

		val leftDist = abs(rect.left - x)
		val topDist = abs(rect.top - y)
		val rightDist = abs(rect.right - x)
		val bottomDist = abs(rect.bottom - y)

		val left: Boolean = leftDist < maxDragDistancePx / viewService.zoom
		val top: Boolean = topDist < maxDragDistancePx / viewService.zoom
		val right: Boolean = rightDist < maxDragDistancePx / viewService.zoom
		val bottom: Boolean = bottomDist < maxDragDistancePx / viewService.zoom

		val xInside = x > rect.left && x < rect.right
		val yInside = y > rect.top && y < rect.bottom

		return when
		{
			xInside && yInside -> MoveType.MOVE
			left && !right && yInside -> MoveType.LEFT_SIDE
			top && !bottom && xInside -> MoveType.TOP_SIDE
			right && !left && yInside -> MoveType.RIGHT_SIDE
			bottom && !top && xInside -> MoveType.BOTTOM_SIDE
			left && top -> MoveType.LEFT_TOP_CORNER
			right && top -> MoveType.RIGHT_TOP_CORNER
			right && bottom -> MoveType.RIGHT_BOTTOM_CORNER
			left && bottom -> MoveType.LEFT_BOTTOM_CORNER
			else -> MoveType.NONE
		}
	}

	override fun onTouchMove(point: PointF, layer: Layer): Boolean
	{
		state = getEditedState(point.toPoint()) ?: return false
		return true
	}

	override fun onTouchStop(point: PointF, layer: Layer): Boolean
	{
		state = getEditedState(point.toPoint()) ?: return false
		return true
	}

	private fun getEditedState(point: Point) = when(val state = state)
	{
		is State.Idle -> null
		is State.Edit.Initial -> state.copy(rect = state.rect.copy(right = snapX(point.x),
		                                                           bottom = snapY(point.y)).sorted())
		is State.Edit.Adjust -> state.copy(rect = getAdjustedRect(point, state))
	}

	private fun getAdjustedRect(point: Point, state: State.Edit.Adjust): Rect
	{
		val deltaX = point.x - state.moveStart.x
		val deltaY = point.y - state.moveStart.y
		return when(state.moveType)
		{
			MoveType.NONE -> state.rectAtStart
			MoveType.LEFT_TOP_CORNER -> state.rectAtStart.copy(left = snapX(state.rectAtStart.left + deltaX),
			                                                   top = snapY(state.rectAtStart.top + deltaY))
			MoveType.RIGHT_TOP_CORNER -> state.rectAtStart.copy(right = snapX(state.rectAtStart.right + deltaX),
			                                                    top = snapY(state.rectAtStart.top + deltaY))
			MoveType.LEFT_BOTTOM_CORNER -> state.rectAtStart.copy(left = snapX(state.rectAtStart.left + deltaX),
			                                                      bottom = snapY(state.rectAtStart.bottom + deltaY))
			MoveType.RIGHT_BOTTOM_CORNER -> state.rectAtStart.copy(right = snapX(state.rectAtStart.right + deltaX),
			                                                       bottom = snapY(state.rectAtStart.bottom + deltaY))
			MoveType.LEFT_SIDE -> state.rectAtStart.copy(left = snapX(state.rectAtStart.left + deltaX))
			MoveType.TOP_SIDE -> state.rectAtStart.copy(top = snapY(state.rectAtStart.top + deltaY))
			MoveType.RIGHT_SIDE -> state.rectAtStart.copy(right = snapX(state.rectAtStart.right + deltaX))
			MoveType.BOTTOM_SIDE -> state.rectAtStart.copy(bottom = snapY(state.rectAtStart.bottom + deltaY))
			MoveType.MOVE ->
			{
				val oldCenter = state.rectAtStart.center()
				val newCenter = snapPoint(oldCenter + Point(deltaX, deltaY))
				state.rectAtStart + (newCenter - oldCenter)
			}
		}.sorted()
	}

	private fun snapX(x: Int) = helpersService.snapX(x.toFloat()).roundToInt()

	private fun snapY(y: Int) = helpersService.snapY(y.toFloat()).roundToInt()

	private fun snapPoint(point: Point) = helpersService.snapPoint(point.toPointF()).toPoint()

	fun applySelection()
	{
		val rect = (state as? State.Edit.Adjust)?.rect ?: return
		val apply: Selection.() -> Selection = when(shape)
		{
			ToolSelectionShape.RECTANGLE -> {{ withRectangleOperation(rect, mode.op) }}
			ToolSelectionShape.OVAL -> {{ withOvalOperation(rect, mode.op) }}
		}
		state = State.Idle
		historyService.commitAction { commit(apply) }
	}

	private fun commit(apply: Selection.() -> Selection): Action.ToRevert = actionPreset.commit {
		val oldSelection = imageService.selection
		imageService.editSelection { apply() }
		toRevert { revert(apply, oldSelection) }
	}

	private fun revert(apply: Selection.() -> Selection, oldSelection: Selection): Action.ToCommit = actionPreset.revert {
		imageService.setSelection(oldSelection)
		toCommit { commit(apply) }
	}

	fun cancelSelection()
	{
		state = State.Idle
	}

	override fun drawOnTop(canvas: Canvas) = canvas.withImageSpace {
		val editState = state as? State.Edit ?: return@withImageSpace
		paint.strokeWidth = selectionLineWidthPx / viewService.zoom
		when(shape)
		{
			ToolSelectionShape.RECTANGLE -> canvas.drawRect(editState.rect, paint)
			ToolSelectionShape.OVAL -> canvas.drawOval(editState.rect.toRectF(), paint)
		}
	}
}
